File: src/main/java/me/zeroeightsix/kami/feature/module/movement/Jesus.java
package me.zeroeightsix.kami.feature.module.movement;

import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.event.KamiEvent;
import me.zeroeightsix.kami.event.events.AddCollisionBoxToListEvent;
import me.zeroeightsix.kami.event.events.PacketEvent;
import me.zeroeightsix.kami.event.events.TickEvent;
import me.zeroeightsix.kami.mixin.client.IPlayerMoveC2SPacket;
import me.zeroeightsix.kami.feature.module.Module;
import me.zeroeightsix.kami.feature.module.Freecam;
import me.zeroeightsix.kami.util.EntityUtil;
import me.zeroeightsix.kami.util.Wrapper;
import net.minecraft.block.FluidBlock;
import net.minecraft.entity.Entity;
import net.minecraft.entity.vehicle.BoatEntity;
import net.minecraft.server.network.packet.PlayerMoveC2SPacket;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Box;
import net.minecraft.util.math.MathHelper;
import net.minecraft.world.EmptyBlockView;

/**
 * Created by 086 on 11/12/2017.
 */
@Module.Info(name = "Jesus", description = "Allows you to walk on water", category = Module.Category.MOVEMENT)
public class Jesus extends Module {

    private static final Box WATER_WALK_AA = new Box(0.D, 0.D, 0.D, 1.D, 0.99D, 1.D);

    @EventHandler
    private Listener<TickEvent.Client> updateListener = new Listener<>(event -> {
        if (!Freecam.INSTANCE.isEnabled()) {
            if (EntityUtil.isInWater(mc.player) && !mc.player.isSneaking()) {
                EntityUtil.updateVelocityY(mc.player, 0.1);
                if (mc.player.getVehicle() != null && !(mc.player.getVehicle() instanceof BoatEntity)) {
                    EntityUtil.updateVelocityY(mc.player.getVehicle(), 0.3);
                }
            }
        }
    });

    @EventHandler
    Listener<AddCollisionBoxToListEvent> addCollisionBoxToListEventListener = new Listener<>((event) -> {
        if (mc.player != null
                && (event.getBlock() instanceof FluidBlock)
                && (EntityUtil.isDrivenByPlayer(event.getEntity()) || event.getEntity()==mc.player)
                && !(event.getEntity() instanceof BoatEntity)
                && !mc.player.isSneaking()
                && mc.player.fallDistance < 3
                && !EntityUtil.isInWater(mc.player)
                && (EntityUtil.isAboveWater(mc.player, false) || EntityUtil.isAboveWater(mc.player.getVehicle(), false))
                && isAboveBlock(mc.player, event.getPos())) {
            Box axisalignedbb = WATER_WALK_AA.offset(event.getPos());
            if (event.getEntityBox().intersects(axisalignedbb)) event.getCollidingBoxes().add(axisalignedbb);
            event.cancel();
        }
    });

    @EventHandler
    Listener<PacketEvent.Send> packetEventSendListener = new Listener<>(event -> {
        if (event.getEra() == KamiEvent.Era.PRE) {
            if (event.getPacket() instanceof PlayerMoveC2SPacket) {
                if (EntityUtil.isAboveWater(mc.player, true) && !EntityUtil.isInWater(mc.player) && !isAboveLand(mc.player)) {
                    int ticks = mc.player.age % 2;
                    if (ticks == 0) {
                        IPlayerMoveC2SPacket xyz = (IPlayerMoveC2SPacket) event.getPacket();
                        xyz.setY(xyz.getY() + 0.02D);
                    }
                }
            }
        }
    });

    @SuppressWarnings("deprecation")
    private static boolean isAboveLand(Entity entity){
        if(entity == null) return false;

        double y = entity.y - 0.01;

        for(int x = MathHelper.floor(entity.x); x < MathHelper.ceil(entity.x); x++)
            for (int z = MathHelper.floor(entity.z); z < MathHelper.ceil(entity.z); z++) {
                BlockPos pos = new BlockPos(x, MathHelper.floor(y), z);

                if (Wrapper.getWorld().getBlockState(pos).getBlock().isFullOpaque(Wrapper.getWorld().getBlockState(pos), EmptyBlockView.INSTANCE, pos)) return true;
            }

        return false;
    }

    private static boolean isAboveBlock(Entity entity, BlockPos pos) {
        return entity.y  >= pos.getY();
    }

}

File: src/main/kotlin/me/zeroeightsix/kami/feature/module/Jesus.kt
package me.zeroeightsix.kami.feature.module

import me.zero.alpine.listener.EventHandler
import me.zero.alpine.listener.EventHook
import me.zero.alpine.listener.Listener
import me.zeroeightsix.kami.event.AddCollisionBoxToListEvent
import me.zeroeightsix.kami.event.KamiEvent
import me.zeroeightsix.kami.event.PacketEvent.Send
import me.zeroeightsix.kami.event.TickEvent.Client.InGame
import me.zeroeightsix.kami.mixin.client.IPlayerMoveC2SPacket
import me.zeroeightsix.kami.util.EntityUtil
import me.zeroeightsix.kami.util.Wrapper
import net.minecraft.block.FluidBlock
import net.minecraft.entity.Entity
import net.minecraft.entity.vehicle.BoatEntity
import net.minecraft.network.packet.c2s.play.PlayerMoveC2SPacket
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Box
import net.minecraft.util.math.MathHelper

@Module.Info(
    name = "Jesus",
    description = "Allows you to walk on water",
    category = Module.Category.MOVEMENT
)
object Jesus : Module() {
    @EventHandler
    private val updateListener = Listener(
        EventHook<InGame> { event: InGame? ->
            if (!Freecam.enabled) {
                if (EntityUtil.isInWater(mc.player) && !mc.player!!.isSneaking) {
                    EntityUtil.updateVelocityY(mc.player, 0.1)
                    if (mc.player!!.vehicle != null && mc.player!!.vehicle !is BoatEntity) {
                        EntityUtil.updateVelocityY(
                            mc.player!!.vehicle,
                            0.3
                        )
                    }
                }
            }
        }
    )

    @EventHandler
    var addCollisionBoxToListEventListener =
        Listener(
            EventHook { event: AddCollisionBoxToListEvent ->
                if (mc.player != null && event.block is FluidBlock
                    && (EntityUtil.isDrivenByPlayer(event.entity) || event.entity === mc.player)
                    && event.entity !is BoatEntity
                    && !mc.player!!.isSneaking
                    && mc.player!!.fallDistance < 3 && !EntityUtil.isInWater(mc.player)
                    && (EntityUtil.isAboveWater(
                        mc.player,
                        false
                    ) || EntityUtil.isAboveWater(
                        mc.player!!.vehicle,
                        false
                    ))
                    && isAboveBlock(
                        mc.player,
                        event.pos
                    )
                ) {
                    val axisalignedbb = WATER_WALK_AA.offset(event.pos)
                    if (event.entityBox.intersects(axisalignedbb)) event.collidingBoxes.add(axisalignedbb)
                    event.cancel()
                }
            }
        )

    @EventHandler
    var packetEventSendListener = Listener(
        EventHook { event: Send ->
            if (event.era === KamiEvent.Era.PRE) {
                if (event.packet is PlayerMoveC2SPacket) {
                    if (EntityUtil.isAboveWater(
                            mc.player,
                            true
                        ) && !EntityUtil.isInWater(mc.player) && !isAboveLand(
                            mc.player
                        )
                    ) {
                        val ticks = mc.player!!.age % 2
                        if (ticks == 0) {
                            val xyz = event.packet as IPlayerMoveC2SPacket
                            xyz.y = xyz.y + 0.02
                        }
                    }
                }
            }
        }
    )

    private val WATER_WALK_AA =
        Box(0.0, 0.0, 0.0, 1.0, 0.99, 1.0)

    private fun isAboveLand(entity: Entity?): Boolean {
        if (entity == null) return false
        val y = entity.y - 0.01
        for (x in MathHelper.floor(entity.x) until MathHelper.ceil(entity.x)) for (z in MathHelper.floor(
            entity.z
        ) until MathHelper.ceil(entity.z)) {
            val pos = BlockPos(x, MathHelper.floor(y), z)

            //if (Wrapper.getWorld().getBlockState(pos).getBlock().isFullOpaque(Wrapper.getWorld().getBlockState(pos), EmptyBlockView.INSTANCE, pos)) return true;
            if (Wrapper.getWorld().getBlockState(pos).isOpaque) return true
        }
        return false
    }

    private fun isAboveBlock(entity: Entity?, pos: BlockPos): Boolean {
        return entity!!.y >= pos.y
    }
}
